libibverbs/verbs.h

```diff
@@ -84,6 +85,54 @@ struct ibv_gid_entry {
 	uint32_t ndev_ifindex;
 };
 
+enum {
+	VERBS_LOG_LEVEL_NONE,
+	VERBS_LOG_ERR,
+	VERBS_LOG_WARN,
+	VERBS_LOG_INFO,
+	VERBS_LOG_DEBUG,
+};
+
+#define verbs_log(ctx, level, format, arg...)                                         \
+do {                                                                                  \
+	if ((level) <= (ctx)->log_level) {                                            \
+		int tmp = errno;                                                      \
+		fprintf((ctx)->log_fp, "%s:%d: " format, __func__, __LINE__, ##arg);  \
+		errno = tmp;                                                          \
+	}                                                                             \
+} while (0)
+
+#define verbs_debug(ctx, format, arg...) \
+	verbs_log(ctx, VERBS_LOG_DEBUG, format, ##arg)
+
+#define verbs_info(ctx, format, arg...) \
+	verbs_log(ctx, VERBS_LOG_INFO, format, ##arg)
+
+#define verbs_warn(ctx, format, arg...) \
+	verbs_log(ctx, VERBS_LOG_WARN, format, ##arg)
+
+#define verbs_err(ctx, format, arg...) \
+	verbs_log(ctx, VERBS_LOG_ERR, format, ##arg)
+
+#ifdef VERBS_DEBUG
+#define verbs_log_datapath(ctx, level, format, arg...) \
+	verbs_log(ctx, level, format, ##arg)
+#else
+#define verbs_log_datapath(ctx, level, format, arg...) {}
+#endif
+
+#define verbs_debug_datapath(ctx, format, arg...) \
+	verbs_log_datapath(ctx, VERBS_LOG_DEBUG, format, ##arg)
+
+#define verbs_info_datapath(ctx, format, arg...) \
+	verbs_log_datapath(ctx, VERBS_LOG_INFO, format, ##arg)
+
+#define verbs_warn_datapath(ctx, format, arg...) \
+	verbs_log_datapath(ctx, VERBS_LOG_WARN, format, ##arg)
+
+#define verbs_err_datapath(ctx, format, arg...) \
+	verbs_log_datapath(ctx, VERBS_LOG_ERR, format, ##arg)
+
```

Comment:
Provider stuff belongs in driver.h, not verbs.h

Reply:
I don't see a reason why libibverbs core function won't use this api as well, should it still move in such case?

Reply:
Yes. It is not public API toward the end user so it does not belong in this header

Reply:
Done.